src/charm.py

```diff
@@ -128,24 +129,12 @@ def _set_status(self, key: Status) -> None:
         log_level: DebugLevel = key.value.log_level
 
         getattr(logger, log_level.lower())(status.message)
-        self.unit.status = status
+        # self.unit.status = status
+        self.unit_statuses.append(key)
 
     def _on_collect_status(self, event: CollectStatusEvent):
-        ready_to_start = self.state.ready_to_start.value.status
-        event.add_status(ready_to_start)
-
-        if not isinstance(ready_to_start, ActiveStatus):
-            return
-
-        if not self.state.runs_broker:
-            # early return, the next checks only concern the broker
-            return
-
-        if not self.broker.workload.active():
-            event.add_status(Status.BROKER_NOT_RUNNING.value.status)
-
-        if not self.state.zookeeper.broker_active():
-            event.add_status(Status.ZK_NOT_CONNECTED.value.status)
+        for status in self.unit_statuses + [Status.ACTIVE]:
+            event.add_status(status.value.status)
```

Comment:
The way the ops internal `collect-status` event handler works is by "sorting" all the statuses by severity and sets the unit / app status to the "highest critical" one, in the order they were passed to `add_status` 
I think this is an interesting design to explore because this behaviour can give us more control over the statuses/errors we display in `juju status`. Like, we could make it so that `self.unit_statuses` is a deck instead of a list and push higher priority statuses to the start of the deck instead of appending at the end.

I am still on the fence about appending `Status.ACTIVE` by default, though. The idea is that if an event handler does not set a "bad" status, we say "ok then, ACTIVE it is" at the end of the charm's execution. I am not sure if we have examples of situations where we want to keep the previous status after the charm execution.

What do you think about this?

Reply:
OK, that makes sense, and thank you for explaining the 'severity' part, I think that's what I was missing.
I think keeping `Status.ACTIVE` default is good, because we can keep adding non-active conditions, but if all our 'checks' pass, we should consider it Active to the best of our knowledge.